src/bidi.js

```diff
@@ -178,6 +182,13 @@ function applyLatinLigatures() {
     });
 }
 
+function applyUnicodeVariationSequences() {
+    const ranges = this.tokenizer.getContextRanges('unicodeVariationSequence');
+    ranges.forEach(range => {
```

Comment:
forEach can be slow due to extra function calls, might be a better idea to use a for-in loop.

Reply:
I personally prefers a more readable code (i.e. forEach) since we can't predict what optimization will appear in the upcoming JS engines...

Reply:
@Connum how many items are expected to be returned (on average) by this this.tokenizer.getContextRanges('unicodeVariationSequence') ?
if more than 100 then I do agree that we should use for() loop.
if lower than 100 then a readable code is more important.

Reply:
Unfortunately, I don't have any empirical values, but from the extent of the lists, I would say that quite a lot could come together if a font were to implement all or at least many of them. But I don't know how likely or common that is. I have no problem changing this to a for loop, in fact I use for loops for opentype.js ever since @ILOVEPIE commeted that, despite using forEach loops everywhere else because I seldomly have real performance concerns when I'm mainly doing trivial web stuff.

Reply:
On the other hand, I copied those lines from another feature handler, so we should probably either change it to for() everywhere, or use forEach() everywhere, especially when we should strave to reduce code duplication in favor of common helper methods to handle se same thing with just a few different parameters.

Reply:
OK, so I wanted to know exactly and I created a benchmark. And I would not have thought that the difference is so significant, even with very few iterations.

Even with one element we are at 2-3%:
https://jsbench.me/rjlet6fft9/1

10 elements - >11%:
https://jsbench.me/rjlet6fft9/2

100 elements - >26%:
https://jsbench.me/rjlet6fft9/3

1000 elements - 91.5% (!):
https://jsbench.me/rjlet6fft9/4

While I do agree that it's a tradeoff with readability/clarity, I also do think that in a library like this we should put more emphasis on performance. So we should probably replace all uses of forEach() with for() - but I would tackle that in a separate PR, and not let it hold this and #535 back, what do you think?

We should also add it to our contribution guidelines (by the way, the PR template references a CONTRIBUTING file, but it seems to have been removed - what happened?) and probably even implement a linting rule so forEach doesn't sneak in in the future. Yes, there may be optimizations in browsers/engines in the future, but as it is now, I think the difference is crucial enough to ban forEach() for the time being.

Reply:
Yeah we can do that as a separate PR, all modern processors do branch prediction/speculative execution to try to counteract how slow function calls are, they're one of the slowest operations the CPU can do.

Reply:
I'm not saying you should write everything in 1 function, but you should try to avoid doing unnecessary function calls in a loop.

Reply:
This is also the reason why compilers will sometimes opt to inline a function call.

Reply:
OK, let's definitely do that then once these PRs are through.